### DISCUSSION TITLE
Is it possible to refresh available options in Datalist on publish/save?

### DISCUSSION THREAD

**Discussion by bielu** [leekelleher/umbraco-contentment#387] [Question]
Hi Lee!
I was wondering if there is by any chance option to refresh editor on save / publish?

**Comment by leekelleher** [comment]
The list-editor UIs are populated with whatever data comes from the server, (at the time of initial content node load), they don't perform their own data fetches. So if Umbraco isn't refreshing the node's property data (and data-type configurations) on save/publish, then the Data List editors wouldn't be refreshed.

Unless there's a way for a property-editor to trigger a refresh that I don't know about? 🤔

**Reply by bielu** [reply]
Hmmm maybe would be better to do API request when opening dialog instead of loading it on inititial content load? same as mntp picker is doing?

**Reply by leekelleher** [reply]
It was an early design decision to populate the Data List editors at content-load time, (I'd envisaged Data Lists to be small/static lists), as opposed to each Data List making their own API fetches. (Fundamentally different from how nuPickers did it). It'd be far too much work involved in re-engineering it now, especially when available dev-time could be spend on the migration to Umbraco v14/Bellissima.

It'd probably be simpler to roll your own custom property-editor if you need a for specific requirement.

**Reply by bielu** [reply]
@leekelleher I am going to make custom editor, but I want still reuse my IDatatlist it is why I asked which service is used, as than I might do in future pr for you so you can adjust it ;)

**Reply by leekelleher** [reply]
Data List does have an API controller that can handle this, it's used for the preview on the data-type configuration.
https://github.com/leekelleher/umbraco-contentment/blob/4.6.1/src/Umbraco.Community.Contentment/DataEditors/DataList/DataListApiController.cs

The `GetPreview()` endpoint needs a `{ alias, dataSource, listEditor}` payload (of the data-type config values), and you'll get back a `{ config, view, alias }` result, which the `config` would contain the `items` for the list-editor.

Hope this helps.

**Reply by leekelleher** [reply]
re: Umbraco v14/Bellissima, I'd be looking at re-engineering how the list items are populated, and load them in via a fetch API endpoint. It'd make it more chatty (over the wire), but offer more context to pass into the API endpoint, e.g. current document, etc.

**Reply by bielu** [reply]
If I will figure out something sensible which can be added as configuration to v5(which can be easily migrated to context api), I will submit pr if not I will keep it as standalone custom editor.

**Reply by bielu** [reply]
Had a look, sadly will need build fully custom solution for it 😢 as it would be intrusive and too complicated to maintain it now in contentment or as exentions which I approach with angular extend 😓

**Reply by leekelleher** [reply]
I'm not sure which type of list-editor UI you want to use, but wondering if the **Data Picker** editor could be a viable alternative? As that uses an API endpoint to fetch the data items when the modal is opened (and supports pagination and querying).

**Reply by bielu** [reply]
I tried to use Data list picker
![image](https://github.com/leekelleher/umbraco-contentment/assets/2244074/cdb0496b-a947-454f-b664-82f3c7a77100)
I might give a try later with Data Picker, but need of implementing search method moved me to working on list picker

**Reply by bielu** [reply]
Data picker actually worked as charm, thanks Lee! 
do you want me to close it or keep open for belissima as option?

**Reply by leekelleher** [reply]
Cool, that's good to hear! Yeah, let's close this one off. 👍
